metricbeat/beater/module.go

```diff
@@ -0,0 +1,204 @@
+package beater
+
+import (
+	"expvar"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/filter"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+)
+
+// Expvar metric names.
+const (
+	successesKey = "success"
+	failuresKey  = "failures"
+	eventsKey    = "events"
+)
+
+var (
+	debugf  = logp.MakeDebug("metricbeat")
+	fetches = expvar.NewMap("fetches")
+)
+
+// moduleWrapper contains the Module and the private data associated with
+// running the Module and its MetricSets. The moduleWrapper contains a list
+// of pointer to its associated MetricSets.
+type moduleWrapper struct {
+	mb.Module
+	filters   *filter.FilterList
+	pubClient publisher.Client
+	children  []*metricSetWrapper
+}
+
+// metricSetWrapper contains the MetricSet and the private data associated with
+// running the MetricSet. It contains a pointer to the parent Module.
+type metricSetWrapper struct {
+	mb.MetricSet
+	parent *moduleWrapper // Parent Module.
+	stats  *expvar.Map    // expvar stats for this MetricSet.
+}
+
+// newModuleWrappers creates new Modules and their associated MetricSets based
+// on the given configuration. It constructs the supporting filters and
+// publisher client and stores it all in a moduleWrapper.
+func newModuleWrappers(
+	modulesConfig []*common.Config,
+	r *mb.Register,
+	publisher *publisher.Publisher,
+) ([]*moduleWrapper, error) {
+	modules, err := mb.NewModules(modulesConfig, r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap the Modules and MetricSet's.
+	var wrappers []*moduleWrapper
+	var errs multierror.Errors
+	for k, v := range modules {
+		debugf("initializing Module type %s, %T=%+v", k.Name(), k, k)
+		f, err := filter.New(k.Config().Filters)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "module %s", k.Name()))
+			continue
+		}
+
+		mw := &moduleWrapper{
+			Module:    k,
+			filters:   f,
+			pubClient: publisher.Connect(),
+		}
+		wrappers = append(wrappers, mw)
+
+		msws := make([]*metricSetWrapper, 0, len(v))
+		for _, ms := range v {
+			debugf("initializing MetricSet type %s/%s, %T=%+v",
+				ms.Module().Name(), ms.Name(), ms, ms)
+			msw := &metricSetWrapper{
+				MetricSet: ms,
+				parent:    mw,
+				stats:     new(expvar.Map).Init(),
+			}
+			msws = append(msws, msw)
+
+			// Initialize expvar stats for this MetricSet.
+			fetches.Set(fmt.Sprintf("%s-%s", mw.Name(), msw.Name()), msw.stats)
+			msw.stats.Add(successesKey, 0)
+			msw.stats.Add(failuresKey, 0)
+			msw.stats.Add(eventsKey, 0)
+		}
+		mw.children = msws
+	}
+
+	return wrappers, errs.Err()
```

Comment:
It feels like we loose some information and decision to drop this information I would think belongs to the caller. But currently we don't need it so lets keep it as it is.